pkg/controllers/clusterthrottle_controller.go

```diff
@@ -378,27 +378,86 @@ func (c *ClusterThrottleController) UnReserveOnClusterThrottle(pod *corev1.Pod,
 func (c *ClusterThrottleController) CheckThrottled(
 	pod *corev1.Pod,
 	isThrottledOnEqual bool,
+	groupNameAnnotation string,
 ) (
 	[]schedulev1alpha1.ClusterThrottle,
 	[]schedulev1alpha1.ClusterThrottle,
 	[]schedulev1alpha1.ClusterThrottle,
 	[]schedulev1alpha1.ClusterThrottle,
+	[]schedulev1alpha1.ClusterThrottle,
 	error,
 ) {
 	throttles, err := c.affectedClusterThrottles(pod)
 	if err != nil {
-		return nil, nil, nil, nil, err
+		return nil, nil, nil, nil, nil, err
 	}
 	affected := []schedulev1alpha1.ClusterThrottle{}
 	alreadyThrottled := []schedulev1alpha1.ClusterThrottle{}
 	insufficient := []schedulev1alpha1.ClusterThrottle{}
+	insufficientIncludingGroup := []schedulev1alpha1.ClusterThrottle{}
 	podRequestsExceedsThreshold := []schedulev1alpha1.ClusterThrottle{}
+
+	// Fetch the pods which have group name's annotation
+	var podGroup []*corev1.Pod
+	if groupNameAnnotation != "" {
+		groupName, isGroup := pod.Annotations[groupNameAnnotation]
+		if isGroup {
+			candidatePods, err := c.podInformer.Lister().Pods(pod.Namespace).List(labels.Everything())
+			if err != nil {
+				return nil, nil, nil, nil, nil, err
+			}
+
+			for _, candidatePod := range candidatePods {
+				if isScheduled(candidatePod) {
+					continue
+				}
+
+				if gn, ok := candidatePod.Annotations[groupNameAnnotation]; !ok || gn != groupName {
+					continue
+				}
+
+				// Don't count the scheduling pod itself.
+				if candidatePod.UID == pod.UID {
+					continue
+				}
+
+				podGroup = append(podGroup, candidatePod)
+			}
+		}
+	}
+
 	for _, thr := range throttles {
 		affected = append(affected, *thr)
 		reservedAmt, reservedPodNNs := c.cache.reservedResourceAmount(types.NamespacedName{Namespace: thr.Namespace, Name: thr.Name})
+
+		usedInGroup := schedulev1alpha1.ResourceAmount{}
+		for _, groupPod := range podGroup {
+			ns, err := c.namespaceInformer.Lister().Get(groupPod.Namespace)
+			if err != nil {
+				return nil, nil, nil, nil, nil, err
+			}
```

Comment:
Good point. I removed the premise that all pods in a group are in the same namespace. This is because this assumption does not make implementation and performance that much better.

Reply:
> This is because this assumption doesn't allow us to improve its implementation and performance that much better than the current implementation in this PR.

Understood.
But, the current implementation assumes all pods in a group are in the same namespace, IIUC.

https://github.com/pfnet/kube-throttler/blob/2f484009943296236f7aa89a7a01d3673e8d8066/pkg/controllers/clusterthrottle_controller.go#L400

We can modify the implementation to fetch the namespace for each pod in a group in the future when we remove this assumption?

Reply:
You are right. I had missed it. Doing so would have meant exploring all the namespace pods, which would have had a huge impact on performance. I still think you are right, let's assume they are in the same namespace.